src/TLuaInterpreter.cpp

```diff
@@ -10739,6 +10750,520 @@ int TLuaInterpreter::restartIrc(lua_State* L)
+        lua_error(L);
+        return 1;
+    }
+
+    speechUnit->setVoice(speechVoices.at(index));
+
+    lua_pushboolean(L, true);
+    return 1;
+}
+
+/**
+ *  Handles playback of queued text.
+ */
+void TLuaInterpreter::ttsStateChanged(QTextToSpeech::State state)
+{
+    TEvent event;
+    switch (state) {
+    case QTextToSpeech::Paused:
+        event.mArgumentList.append(QLatin1String("ttsSpeechPaused"));
+        break;
+    case QTextToSpeech::Speaking:
+        event.mArgumentList.append(QLatin1String("ttsSpeechStarted"));
+        break;
+    case QTextToSpeech::BackendError:
+        event.mArgumentList.append(QLatin1String("ttsSpeechError"));
+        break;
+    case QTextToSpeech::Ready:
+        event.mArgumentList.append(QLatin1String("ttsSpeechReady"));
+        break;
+    }
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+
+    if (state == QTextToSpeech::Speaking) {
+        event.mArgumentList.append(speechCurrent);
+        event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    }
+
+    mudlet::self()->getHostManager().postInterHostEvent(NULL, event, true);
+
+    if (state != QTextToSpeech::Ready || speechQueue.empty()) {
+        return;
+    }
+
+    QString textToSay;
+    textToSay = speechQueue.takeFirst();
+
+    speechUnit->say(textToSay);
+    speechCurrent = textToSay;
+
+    return;
+}
+
+
+/** ttsQueueSpeech( input, opt. index )
+ *  Adds a line of text to the queue at the specified index.
+ *  If no index is provided, it will add it to the end of the queue.
+ */
+int TLuaInterpreter::ttsQueueSpeech(lua_State* L)
+{
+    TLuaInterpreter::ttsBuild();
+
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "ttsQueueText: bad argument #1 type (input as string expected, got %s!)", luaL_typename(L, 1));
+        lua_error(L);
+        return 1;
+    }
+
+    QString inputText = lua_tostring(L, 1);
+    int index;
+
+    if (lua_gettop(L) > 1) {
+        if (!lua_isnumber(L, 2)) {
+            lua_pushfstring(L, "ttsQueueText: bad argument #2 type (optional index as number expected, got %s!)", luaL_typename(L, 1));
+            lua_error(L);
+            return 1;
+        }
+
+        index = lua_tonumber(L, 2);
+        index--;
+
+        if (index < 0) {
+            index = 0;
+        }
+
+        if (index > speechQueue.size()) {
+            index = speechQueue.size();
+        }
+    } else {
+        index = speechQueue.size();
+    }
+
+    speechQueue.insert(index, inputText);
+
+    TEvent event;
+    Host& host = getHostFromLua(L);
+    event.mArgumentList.append(QLatin1String("ttsSpeechQueued"));
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    event.mArgumentList.append(inputText);
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    event.mArgumentList.append(QString::number(index));
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_NUMBER);
+    host.raiseEvent(event);
+
+    TLuaInterpreter::ttsStateChanged(speechUnit->state());
```

Comment:
I don't get the logic here... When nothing plays, we enter the function with the state `Ready`. That means we raise this event to all profiles (which may start playing/queueing their own TTS stuff) but then continue to start playing TTS things ourselves.

When we are already playing TTS things, we enter with `Speaking`, which in turn comes out as `Started`, even though we may have started some time ago and may raise the event multiple times in a row for each queued thing. I'd expect a "running" event instead and "started" only once as long as I didn't receive a "stop".

Reply:
That event is new text being queued for speech, not anything to do with the state.

The state is called there in case the queue is empty, to signify to the ttsStateChanged function that it's ready to start synthesizing the current queue, instead of waiting for a ttsSpeak or some other such. That's the only place ttsStateChanged is manually called, everything else (including possible quirks with started vs ready) is pulled directly from the QT connected event, not the logic of the queue.

Reply:
As well, since I can't reply to the above note about ttsSpeak() for some reason, it interrupts the current speech (not clearing the queue, just interrupting the current line). All ttsSpeechChanged events are raised by the connected QT event (except for the one exception mentioned above), so no need to add those in there.